docs/rfcs/0008-simplify-datastore.md

```diff
@@ -0,0 +1,126 @@
+    - In some cases, when the lock isn't released successfully, such as owner/writer faulted before free the lock on objects, we have to implement timeout or such -> increase the complexity and maintain cost to the system.
+
+#### <3> Using version field as data object properties (optimistic locking)
+
+Use a version field in record to ensure that correct version was used before writing. 
+
+Pros:
+
+    - Simple to implement since we only need to verify the version whenever we need to write.
+
+Cons:
+
+    - Depends on which file store being used, the filestore has to support sync state between regions/environment. Currently, only GSC is supporting features to implement this.
+
+### Achieve by avoid concurrent processes data as much as possible
+
+The idea is to split data to ensure that each file is only owned by one writer
+
+![image](assets/simplify-datastore-base-idea.png)
+
+Pros:
+
+    - Simple to implement, since PipeCD data model is simple, we prepared for single writer already.
+
+Cons:
+
+    - Must keep a single writer for each object (or file in case we separate object to multiple files) and narrow down the number of possible features we can do
+
+To handle (2), there are some things we have to be aware of
+- Filter option interface <1>
+- Query bulk (fetch list) and quere by id (fetch single object) <2>
+- Read complex object which required to fetch multiple files then build up before return <3>
+
+#### <1> Filter option interface
+
+To unify with other datastore which be supported as PipeCD datastore (GCP firestore and MySQL), we plan to keep the ListOptions interface. Besides, this change should only affect the datastore layer, and we can change the application/API layer a bit to adopt this filestore database, but should not be too much. With this kind of requirement, we will have to add logic to the filestore database, which will fetch all objects in pagination and then filter by the conditions requested by clients (pipectl or web or piped). This may affect the performance and API latency compared to the database used pattern, but we can reduce the affection by using Cache.
+
+#### <2> Query bulk (fetch list) and quere by id (fetch single object)
+
+Since we will use filestore as PipeCD database, we should implement filestore with hot & cold storage separated pattern - storage called hot or cold defined based on: How often do the objects be updated? or the update frequency.
+
+Objects such as: deployment, command, event, deployment-chain, etc are only be created once and once they finished their life time (state or status changed to SUCCESS|FAILURE|CANCELLED), they will no more be updated. Those kinds of objects should be stored in `chunk` and mostly will be used to shown as list from deployment list view, for example. 
+
+On the opposite side, objects like pipelines, projects, applications, etc. have fewer numbers and are updated frequently throughout their lifetime; they should therefore be stored in hot storage.
```

Comment:
`pipelines`? w

Reply:
Nice catch; it should be `piped`. My autosuggestion tool changed that without my permission 🤣

Reply:
Fixed, thank you 🙏